src/core/MIGRATION.md

```diff
@@ -450,8 +450,13 @@ We now move this logic into a new plugin definition, which is based off of the c
 // server/plugin.ts
 import { ElasticsearchPlugin } from '../elasticsearch';
 
-interface CoreSetup {
+// note: We use a name unique to our plugin for this type since our shimmed interface is not 100%
+// compatible with NP's CoreSetup.
+interface DemoPluginCoreSetup {
   elasticsearch: ElasticsearchPlugin // note: Elasticsearch is in Core in NP, rather than a plugin
+  http: {
+    route: Legacy.Server['route']    // note: NP uses `http.createRouter()`
```

Comment:
Is there any intention to provide `core.http.route` as a method in the NP, that works like this? If not, then I think showing this as a "shim" is confusing because I interpret "shim" as "massage the old way so that it _looks_ and _works_ like the new way".

Reply:
In other words if in NP `core.http` exists, but `core.http.route` isn't going to exist, than this shim breaks your future code in a hard to notice way. 

If we need an example of how to keep using legacy router in the shim (do we need this?) I would say maybe shimming it as `core.legacyHttp.route` is a clearer way of indicating what's going on.

Reply:
I like that idea a lot to be honest.

Reply:
Another option is to use the `__legacy` namespace that some plugins have been doing:

```ts
import { CoreSetup } from 'src/core/server';

export interface DemoPluginCoreSetup extends CoreSetup {
  __legacy: {
    server: {
      route: Legacy.Server['route'];
    };
  };
}

export class Plugin {
  public setup(core: CoreSetup, plugins: {}) {
    core.__legacy.route({...})
  }
}
```

Reply:
I like what you have here and would be happy to see it merge.

> core.http.route will never exist.

Is that correct? I opened https://github.com/elastic/kibana/issues/44174 about `core.http.route` and @restrry [replied](https://github.com/elastic/kibana/issues/44174#issuecomment-528764696) "should be done as a part of #44620" so I thought it was coming eventually

Reply:
We have broad consensus that registering routes with a configuration object could be a nicer API but I don't think we have concrete designs for how it will be exposed. I imagined that we might still have a `core.http.createRouter(path: string)` method like today, but that the Router instance returned from this would allow you to do `router.add({...})` or `router.register({...})`

However, this conversation kinda changed my mind on the our approach to shimming plugins. We used to recommend constructing a `CoreSetup` that looks as close as possible to what CoreSetup will look like in the New Platform. This has caused a lot of confusion and speculation which I realise now doesn't add much value.

As @jasonrhodes pointed out:
> then this shim breaks your future code in a hard to notice way.

Instead of making it look the _same_, I think the goal should rather be to have legacy be as different and noticeable as possible, and make the exact legacy API surface explicit by picking only the legacy methods that you depend on.

Reply:
Exposing a function on a router instance, vs `core.http` is also what I was expecting. Thanks for the explanation & confirmation. Sorry for the noise.